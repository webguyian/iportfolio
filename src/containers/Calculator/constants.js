export const add = '+';
export const allClear = 'AC';
export const backspace = 'Backspace';
export const clear = 'C';
export const decimal = '.';
export const divide = '/';
export const divideAlt = '\u00f7';
export const enter = 'Enter';
export const equals = '=';
export const multiply = '*';
export const multiplyAlt = '\u00d7';
export const percent = '%';
export const positiveNegative = '+/-';
export const subtract = '-';
export const zero = '0';

export const controls = {
  add,
  allClear,
  clear,
  decimal,
  divide,
  divideAlt,
  equals,
  multiply,
  multiplyAlt,
  percent,
  positiveNegative,
  subtract
};

export const allOperators = [
  add,
  allClear,
  clear,
  decimal,
  divide,
  divideAlt,
  equals,
  multiply,
  multiplyAlt,
  percent,
  positiveNegative,
  subtract
];

export const operators = [
  add,
  allClear,
  clear,
  divide,
  divideAlt,
  multiply,
  multiplyAlt,
  positiveNegative,
  subtract
];

export const controlLabels = {
  [allClear]: 'All clear',
  [clear]: 'Clear',
  [positiveNegative]: 'Positive/negative',
  [percent]: 'Percent',
  [multiplyAlt]: 'Multiply',
  [divideAlt]: 'Divide',
  [add]: 'Add',
  [subtract]: 'Subtract',
  [equals]: 'Equals',
  [decimal]: 'Decimal'
};

export const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

export const keys = {
  backspace,
  enter
};

export const operatorKeys = [backspace, enter]
  .concat(...allOperators)
  .concat(numbers);
