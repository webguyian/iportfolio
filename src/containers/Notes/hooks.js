import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
