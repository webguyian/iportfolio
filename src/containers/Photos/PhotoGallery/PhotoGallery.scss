.photo-gallery {
  @include stretch-column-nowrap;
  max-height: 700px;
  overflow-y: auto;

  > .ui-text--display {
    @include type-title;
  }

  &-grid {
    display: grid;
    padding: 0 $gutter;
    grid-template-columns: repeat(3, auto);
    grid-gap: $gutter;
  }

  &-photo-frame {
    max-width: 185px;
    animation: loading-scale-in 0.4s cubic-bezier(0.39, 0.58, 0.57, 1);

    &--favorite {
      @include type-icon("\f004");
      max-width: 185px;
      animation: loading-scale-in 0.4s cubic-bezier(0.39, 0.58, 0.57, 1);

      &::after {
        position: absolute;
        bottom: 8px;
        left: 8px;
      }
    }
  }

  &-photo {
    max-width: 100%;
  }

  &-empty {
    margin: auto $gutter-m;
  }

  &-empty-message {
    margin-bottom: $gutter;
    font-size: $font-size-l;
    text-align: center;
  }
}

@keyframes loading-scale-in {
  from {
    transform: scale(0.25);
  }
}
