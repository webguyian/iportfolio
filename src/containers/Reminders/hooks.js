import { useEffect, useState, useRef } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useLocalStorage = (key, initialValue) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error return initialValue
      return initialValue;
    }
  };
  const [storedValue, setStoredValue] = useState(getValue);
  const setValue = value => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;

      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      // Drop error
    }
  };

  return [storedValue, setValue, getValue];
};

export const useRefFocus = (node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element) {
      // Add focus to element
      element.focus();
    }
  }, [ref]);

  return ref;
};

export const useRefControlledFocus = (focused, node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element && focused) {
      element.focus();
    }
  }, [focused]);

  return ref;
};

export const useReminders = (reminders, setReminders) => {
  const remindersRef = useRef(reminders);

  useEffect(() => {
    // Update reminders in ref
    remindersRef.current = reminders;
  });

  useEffect(() => {
    return () => {
      const { current } = remindersRef;
      const updatedReminders = current.filter(reminder =>
        reminder.value.trim()
      );

      // Trigger callback without empty reminders
      setReminders(updatedReminders);
    };
  }, [remindersRef]);

  return remindersRef;
};

export const useSwipeOffset = (callback, initial = 0) => {
  const [offset, setOffset] = useState(initial);

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Left') {
      // Exit early if not left swipe
      return false;
    }

    setOffset(eventInfo.deltaX);
  };

  const handleSwipeLeft = eventInfo => {
    const { deltaX } = eventInfo;

    if (deltaX > 80) {
      callback(eventInfo);
    }
  };

  const handleSwipeRight = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedLeft: handleSwipeLeft,
    onSwipedRight: handleSwipeRight,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset, setOffset];
};
