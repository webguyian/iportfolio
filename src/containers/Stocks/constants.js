export const initialStocks = [
  { symbol: 'AAPL', displaySymbol: 'AAPL', description: 'APPLE INC' },
  { symbol: 'AMZN', displaySymbol: 'AMZN', description: 'AMAZON.COM INC' },
  { symbol: 'FB', displaySymbol: 'FB', description: 'FACEBOOK INC-CLASS A' },
  { symbol: 'GOOG', displaySymbol: 'GOOG', description: 'ALPHABET INC-CL C' },
  { symbol: 'MSFT', displaySymbol: 'MSFT', description: 'MICROSOFT CORP' }
];

export const API_CANDLESTICK = 'https://finnhub.io/api/v1/stock/candle';

export const API_NEWS = 'https://finnhub.io/api/v1/news';

export const API_QUOTE = 'https://finnhub.io/api/v1/quote';

export const API_TOKEN = 'bojsah7rh5rcji5m4udg';

export const API_EXCHANGE = `https://finnhub.io/api/v1/stock/symbol?exchange=US&token=${API_TOKEN}`;
