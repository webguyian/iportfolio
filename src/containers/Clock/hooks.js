import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
