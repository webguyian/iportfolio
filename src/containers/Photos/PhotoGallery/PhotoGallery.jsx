import React from 'react';
import { bool, object } from 'prop-types';
import cx from 'classnames';

import Link from 'components/Link/Link';
import Text from 'components/Text/Text';
import PhotoView from 'containers/Photos/PhotoView/PhotoView';

import { formatCreatedDate, getPhotoIndex } from 'modules/photos/helpers';
import { usePhotos } from 'modules/photos/hooks';

const PhotoGallery = props => {
  const { favoritesOnly, location } = props;
  const [allPhotos, , favorites] = usePhotos();
  const photoIndex = getPhotoIndex(location);
  const photos = favoritesOnly ? favorites : allPhotos;
  const currentPhoto = photos[photoIndex];
  const pathType = favoritesOnly ? 'favorites' : 'gallery';
  const title = favoritesOnly ? 'Favorites' : 'Photos';
  const message = favoritesOnly
    ? 'Tap the heart to favorite some of your photos.'
    : 'Take snapshots from the Camera app.';
  const baseClass = 'photo-gallery';
  const photoClass = `${baseClass}-photo`;
  const frameClass = `${photoClass}-frame`;
  const favoriteClass = `${frameClass}--favorite`;

  if (currentPhoto) {
    return <PhotoView photo={currentPhoto} />;
  }

  return (
    <div className={baseClass}>
      <Text element="h1" type="display">
        {title}
      </Text>
      {photos && photos.length ? (
        <div className={`${baseClass}-grid`}>
          {photos.map((photo, index) => (
            <Link
              key={photo.metadata.dateCreated}
              className={cx(
                frameClass,
                photo.metadata.favorited && favoriteClass
              )}
              to={`/photos/${pathType}/${index}`}
              state={{ theme: 'black-dark' }}
            >
              <img
                className={photoClass}
                alt={`Photo taken ${formatCreatedDate(photo)}`}
                src={photo.image}
              />
            </Link>
          ))}
        </div>
      ) : (
        <div className={`${baseClass}-empty`}>
          <Text
            className={`${baseClass}-empty-message`}
            element="p"
            modifier="bold"
          >
            No {title.toLowerCase()} found.
          </Text>
          <Text className={`${baseClass}-empty-message`} element="p">
            {message}
          </Text>
        </div>
      )}
    </div>
  );
};

PhotoGallery.propTypes = {
  favoritesOnly: bool,
  location: object.isRequired
};

PhotoGallery.defaultProps = {
  favoritesOnly: false
};

export default PhotoGallery;
