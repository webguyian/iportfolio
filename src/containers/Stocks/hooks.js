import { useEffect, useState } from 'react';

import { useStorageCache } from 'hooks';
import { API_EXCHANGE, API_TOKEN } from 'containers/Stocks/constants';
import {
  getCandlestickEndpoint,
  getCandlestickEndpoints,
  getStockEndpoint,
  getStockNewsEndpoint,
  isExpired,
  isExpiredNews,
  updateChartData,
  rehydrateChartData
} from 'containers/Stocks/helpers';

export const useFetch = endpoint => {
  const [data, setData] = useState(null);
  const fetchData = async () => {
    const response = await fetch(endpoint);
    const result = await response.json();

    setData(result);
  };

  useEffect(() => {
    if (!endpoint) {
      return;
    }

    fetchData();
  }, [endpoint]);

  return data;
};

export const useFetchAll = urls => {
  const [responses, setResponses] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const results = await Promise.all(
        urls.map(async url => {
          const response = await fetch(url);

          return response.json();
        })
      );

      setResponses(results);
    };

    if (urls.length) {
      fetchData();
    }
  }, [urls]);

  return responses;
};

export const useSearch = () => {
  const [value, setValue] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const [searchFocus, setSearchFocus] = useState(false);
  const hasSearch = Boolean(searchFocus || searchTerm.length);

  const handleCancel = () => {
    setValue('');
    setSearchTerm('');
    setSearchFocus(false);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    setSearchFocus(focus => !focus);
  };

  const searchHandlers = {
    onBlur: handleFocus,
    onCancel: handleCancel,
    onChange: handleChange,
    onFocus: handleFocus,
    value
  };

  useEffect(() => {
    if (value) {
      setSearchTerm(value.toUpperCase().trim());
    } else {
      setSearchTerm('');
    }
  }, [value]);

  return [searchTerm, searchHandlers, hasSearch];
};

export const useStockSearch = (allStocks, searchTerm) => {
  const [filteredStocks, setFilteredStocks] = useState(allStocks);

  useEffect(() => {
    if (searchTerm.length > 1) {
      const filtered = allStocks.filter(stock => {
        return (
          stock.symbol.indexOf(searchTerm) === 0 ||
          stock.description.indexOf(searchTerm) === 0
        );
      });

      setFilteredStocks(filtered);
    } else if (!searchTerm) {
      setFilteredStocks(allStocks);
    }
  }, [searchTerm]);

  return filteredStocks;
};

export const useStocks = initialStocks => {
  const [endpoints, setEndpoints] = useState([]);
  const [stocks, setStocks] = useState([]);
  const [allStocks, setAllStocks] = useState([]);
  const [activeStocks, setActiveStocks] = useState([]);
  const exchangeResponse = useFetch(API_EXCHANGE);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stocks',
    { stocks },
    stockData => !stockData.stocks.length
  );

  useEffect(() => {
    const cachedStocks = cache && cache.stocks;

    if (cachedStocks && cachedStocks.length) {
      setStocks(cachedStocks);
      setActiveStocks(cachedStocks);
    } else {
      setStocks(initialStocks);
      setActiveStocks(initialStocks);
    }
  }, []);

  useEffect(() => {
    if (!exchangeResponse) {
      // Exit early if no response
      return;
    }

    const symbols = activeStocks.map(stock => stock.symbol);
    const currentStocks = exchangeResponse.filter(stock =>
      symbols.includes(stock.symbol)
    );

    setEndpoints(currentStocks.map(getStockEndpoint));
    setAllStocks(exchangeResponse);
  }, [exchangeResponse]);

  useEffect(() => {
    if (activeStocks.length) {
      setEndpoints(activeStocks.map(getStockEndpoint));
    }
  }, [activeStocks]);

  useEffect(() => {
    if (!response.length || !activeStocks.length) {
      return;
    }

    const updatedStocks = response.map((result, index) => {
      const stock = activeStocks[index];

      return {
        ...result,
        ...stock,
        price: result.c,
        previousPrice: result.pc
      };
    });

    setStocks(updatedStocks);
  }, [response, activeStocks]);

  return [stocks, setStocks, allStocks];
};

export const useStockDetail = () => {
  const [endpoint, setEndpoint] = useState('');
  const [activeStock, setActiveStock] = useState(null);
  const [stock, setStockDetail] = useState(activeStock);
  const response = useFetch(endpoint);
  const getStock = symbol =>
    `https://finnhub.io/api/v1/quote?symbol=${symbol}&token=${API_TOKEN}`;

  useEffect(() => {
    setStockDetail(activeStock);

    if (!activeStock) {
      return;
    }

    if (!activeStock.price) {
      setEndpoint(getStock(activeStock.symbol));
    }
  }, [activeStock]);

  useEffect(() => {
    if (response) {
      setStockDetail({
        ...activeStock,
        ...response,
        price: response.c,
        previousPrice: response.pc
      });
      setEndpoint('');
    }
  }, [response]);

  return [stock, setActiveStock];
};

export const useStockNews = category => {
  const [endpoint, setEndpoint] = useState('');
  const [news, setNews] = useState([]);
  const response = useFetch(endpoint);
  const cache = useStorageCache(
    'stock-news',
    { news: response },
    res => !res.news.length
  );

  useEffect(() => {
    const cachedNews = cache && cache.news;

    if (cachedNews && cachedNews.length) {
      setNews(cachedNews);
      const expired = isExpiredNews(cachedNews[0].datetime);

      if (!expired && !category) {
        // Exit early when cache is not expired and no category
        return;
      }
    }

    setEndpoint(getStockNewsEndpoint(category));
  }, [category]);

  useEffect(() => {
    if (response && response.length) {
      setNews(response);
    }
  }, [response]);

  return news;
};

export const useStockChart = (symbol, range) => {
  const [endpoint, setEndpoint] = useState();
  const [updatedResponse, setResponse] = useState(null);
  const response = useFetch(endpoint);

  useEffect(() => {
    const updatedEndpoint = getCandlestickEndpoint(symbol, range);

    setEndpoint(updatedEndpoint);
  }, [symbol, range]);

  useEffect(() => {
    if (response && response.s === 'ok') {
      const data = updateChartData(response, range);

      setResponse(data);
    }
  }, [response]);

  return updatedResponse;
};

export const useStockTicker = (stocks, range = '1D') => {
  const [endpoints, setEndpoints] = useState([]);
  const [updatedResponse, setResponse] = useState([]);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stock-charts',
    { charts: updatedResponse },
    res => !res.charts.length
  );

  useEffect(() => {
    const cachedCharts = cache && cache.charts;

    if (cachedCharts && cachedCharts.length) {
      const updated = cachedCharts.map(rehydrateChartData);
      const [head] = updated;
      const expired = isExpired(head.chartData.endDate);

      setResponse(updated);

      if (!expired) {
        // Exit early when cache is not expired
        return;
      }
    }

    // Trigger fetch for candlestick data
    setEndpoints(getCandlestickEndpoints(stocks, range));
  }, [stocks.length]);

  useEffect(() => {
    if (response.length) {
      const updated = response.map((res, index) => {
        if (res.s !== 'ok') {
          return null;
        }

        const stock = stocks[index];
        const chartData = updateChartData(res, range);

        return {
          ...stock,
          chartData
        };
      });

      if (updated[0] !== null) {
        setResponse(updated);
      }
    }
  }, [response]);

  return updatedResponse;
};
