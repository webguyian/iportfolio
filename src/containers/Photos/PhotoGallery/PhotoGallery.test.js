import React from 'react';
import { render } from '@testing-library/react';

import { mockPhoto } from 'modules/photos/constants';
import * as hooks from 'modules/photos/hooks';

import PhotoGallery from './PhotoGallery';

describe('<PhotoGallery />', () => {
  const props = { location: { pathname: '/photos/gallery' } };
  const usePhotos = hooks.usePhotos;

  hooks.usePhotos = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.usePhotos.mockImplementation(usePhotos);
  });

  it('renders correctly', () => {
    const { asFragment } = render(<PhotoGallery {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with photos', () => {
    const photos = [mockPhoto];

    hooks.usePhotos.mockReturnValueOnce([photos]);
    const { asFragment } = render(<PhotoGallery {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with current photo', () => {
    const photos = [mockPhoto];

    hooks.usePhotos.mockReturnValueOnce([photos]);
    const { asFragment } = render(
      <PhotoGallery location={{ pathname: '/photos/gallery/0' }} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
