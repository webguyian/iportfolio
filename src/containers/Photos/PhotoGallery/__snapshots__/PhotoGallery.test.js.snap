// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoGallery /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="photo-gallery"
  >
    <h1
      class="ui-text ui-text--display"
    >
      Photos
    </h1>
    <div
      class="photo-gallery-empty"
    >
      <p
        class="ui-text ui-text--body ui-text--bold photo-gallery-empty-message"
      >
        No photos found.
      </p>
      <p
        class="ui-text ui-text--body photo-gallery-empty-message"
      >
        Take snapshots from the Camera app.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<PhotoGallery /> renders correctly with current photo 1`] = `<DocumentFragment />`;

exports[`<PhotoGallery /> renders correctly with photos 1`] = `
<DocumentFragment>
  <div
    class="photo-gallery"
  >
    <h1
      class="ui-text ui-text--display"
    >
      Photos
    </h1>
    <div
      class="photo-gallery-grid"
    >
      <mock-link
        classname="photo-gallery-photo-frame photo-gallery-photo-frame--favorite"
        state="[object Object]"
        to="/photos/gallery/0"
      >
        <img
          alt="Photo taken Oct 1, 2019 at 11:11 AM"
          class="photo-gallery-photo"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII="
        />
      </mock-link>
    </div>
  </div>
</DocumentFragment>
`;
